### ISSUE TITLE
Ray/Velocity Direction in Pure Cannon.js

### ISSUE THREAD

**Issue by RaggarDK** [schteppe/cannon.js#305]
Okay, if we take the three.js FPS demo as an example, it uses the camera to get the direction in which it applies the velocities. Let's assume that I want the same functionality, but without having a camera object.
The reason for this, is that I want the code to run purely on node.js without three.js or babylon.js. This is part of a server-client implementation.
I mention the FPS example because the shooting mechanics is what I'm looking to implement.
I would prefer being able to use a ray, but a projectile-like body would suffice, too.
I'm just not sure as to how this can be implemented.

I made an example here, using babylon.js:
http://www.babylonjs-playground.com/#15U9CT#3

Using the Cannon.js code from this example, I am able to move the sphere around using the WASD keys, as well as rotate the object using the mouse. Now, the keyboard input and rotation is sent to the node.js server using sockets, and this seems to function without too many issues. I know that moving the position instead of applying velocity is a bad idea, but I have no idea how to find the proper velocity values based on the rotation, so that the sphere always moves in the direction of the rotation/mouse position.

So without the camera, and only with the body(position/rotation/quaternion), how would I cast a ray straight out the front of the spheres position and rotation, to check for intersects against other objects?

**Comment by schteppe** [comment]
Hi,
If you want to get a local direction vector from a Cannon.js body in world space, you can use `Body.prototype.vectorToWorldFrame`.

``` js
var localForward = new CANNON.Vec3(0,0,-1); // correct?
var worldForward = new CANNON.Vec3();
body.vectorToWorldFrame(localForward, worldForward);
```

Now that you have the forward direction, you can use it to do your raycasting. Here's a demo on how to use the raycasting API in cannon.js: https://github.com/schteppe/cannon.js/blob/master/demos/trimesh.html

Unfortunately I don't have time to give you a full body-ro-ray code at this moment but if you need it I could probably do it later... It's just some vector math really.

**Comment by RaggarDK** [comment]
Hey. Thanks for your reply. And awesome library ;)
I'm not entirely sure as to how that should be implemented.

And when it comes to the ray, the position of the rigidbody is not a problem, but how do I obtain the second vector, in a straight line in the direction the body is facing, based on the rotation?

This might be simple, but I really can't wrap my head around it.

**Comment by schteppe** [comment]
That's what the code I provided does :) Rotates the local forward vector according to the body rotation.. maybe I misunderstood what you need?

**Comment by RaggarDK** [comment]
Nah. I'm pretty sure I'm the one who doesn't quite get it.
Maybe if understand the solution to the directionsal problem, it might be easier to get the second issue fixed as well.
Let's take this very basic example, using velocity instead of positioning: 
http://www.babylonjs-playground.com/#MBQIR

If you move the mouse around, the sphereBody rotates accordingly.
If you press WASD, the sphere moves straight left, forward, etc. not taking into account the direction that it is now facing. This was fixed in my first linked example, but I'm pretty sure that was a bad hack, as I moved the position, instead of changing the velocity.

I'm not sure how I can do the same as the Three.js FPS example, as that uses empty objects, whereas I use the body itself in order for the code to be run server side using just node.

If I apply the following code in my loop:
 var localForward = new CANNON.Vec3(0,0,1); 
        var worldForward = new CANNON.Vec3();
        sphereBody.vectorToWorldFrame(localForward, worldForward);
         console.log(worldForward);
I get results like this depending on where I move to:
Object { x: 0.3476713759012254, y: -0.7371399860171481, z: 0.5794387417092405 

And if I log the position of the sphereBody, it is of course different to above vector.

So in other words, how do I take the rotation of the sphereBody, and apply this in such a way, that if I press W, the velocity.z is based on where the sphereBody is facing. Same goes for x of course.

And this should of course be done using cannon.js without third part objects and cameras.
I suppose it is possible.

**Comment by schteppe** [comment]
Ok, something like this?

``` js
var speed = 10;
var velocity = new CANNON.Vec3();
worldForward.y = 0; // don't need up velocity, so clamp it
worldForward.normalize();
worldForward.scale(speed, velocity);
console.log(velocity);
```

**Comment by RaggarDK** [comment]
That gives me a value (only z) of between -10 and 10.
10 for the first second, so the body moves forward when applying the velocity, but then start being between the two, making it impossible to move.

Can't you like reset the quaternion with new rotational data? I feel completely stupid right now.

I'll take a look at the fps demo and see if I can recreate it. Somehow that demo applies the velocity in new directions according to the camera.

**Comment by RaggarDK** [comment]
I reverted back to Three.js for now.
Seems like I need some more time to adjust to Babylon.

**Comment by GulfBull** [comment]
Hello Raggar.
If you find a way of solving the ray issue, please let me know.
I am using three and too trying to determine the 2nd position of the ray based on the quaternion of my body. The snippet posted by schteppe seems inconsistent in my tests.

**Comment by GulfBull** [comment]
By inconsistent I mean that by using the piece by Schteppe
`var localForward = new CANNON.Vec3(0,0,-1); // correct?
    var worldForward = new CANNON.Vec3();
    sphereBody.vectorToWorldFrame(localForward, worldForward);
    console.log(worldForward);`
simply in the click eventhandler. Whenever a click is registered the Z and X values of the vector3 changes drastically without any movement or rotation. Howver the Y value stays the same.

Look at the following code added in the very same eventhandler in the three fps demo
`                    console.log(worldForward);

```
                var vector = new THREE.Vector3(sphereBody.position.x + 2, sphereBody.position.y + 2, sphereBody.position.z + 1.5);
                var result2 = new THREE.Vector3();
                var to = result2.sub(sphereBody.position).normalize();
                var result = new CANNON.RaycastResult();
                var raycastOptions = {};
                result.reset();
                world.raycastClosest(vector, to, raycastOptions, result);
                testBallMesh.position.copy(result.hitPointWorld);
                console.log(testBallMesh.position);
                console.log(result.shape);`
```

If I instead use the already calculated vector (shootDirection) from the getShootDir function it successfully intersects with the ground plane when aimed at it. But randomly with the box shapes aswell.

The intersection point and obviously also the testBallMesh is at all time positioned within the -1 to 1 range.
